libs/MobileSync/src/com/salesforce/androidsdk/mobilesync/target/SyncDownTarget.kt

```diff
@@ -0,0 +1,375 @@
+        var additionalPredicate = ""
+        val indexSpecs = syncManager.smartStore.getSoupIndexSpecs(soupName)
+        for (indexSpec in indexSpecs) {
+            if (indexSpec.path == SYNC_ID) {
+                additionalPredicate = String.format(
+                    Locale.US,
+                    "AND {%s:%s} = %d",
+                    soupName,
+                    SYNC_ID,
+                    syncId
+                )
+                break
+            }
+        }
+        return additionalPredicate
+    }
+
+    /**
+     * Return ids of non-dirty records (records NOT locally created/updated or deleted)
+     * @param syncManager
+     * @param soupName
+     * @param idField
+     * @param additionalPredicate
+     * @return
+     * @throws JSONException
+     */
+    @Throws(JSONException::class)
+    fun getNonDirtyRecordIds(
+        syncManager: SyncManager, soupName: String,
+        idField: String, additionalPredicate: String
+    ): SortedSet<String> {
+        val nonDirtyRecordsSql = getNonDirtyRecordIdsSql(soupName, idField, additionalPredicate)
+        return getIdsWithQuery(syncManager, nonDirtyRecordsSql)
+    }
+
+    /**
+     * Return SmartSQL to identify non-dirty records
+     * @param soupName
+     * @param idField
+     * @param additionalPredicate
+     * @return
+     */
+    protected open fun getNonDirtyRecordIdsSql(
+        soupName: String,
+        idField: String,
+        additionalPredicate: String
+    ): String {
+        return String.format(
+            "SELECT {%s:%s} FROM {%s} WHERE {%s:%s} = 'false' %s ORDER BY {%s:%s} ASC",
+            soupName,
+            idField,
+            soupName,
+            soupName,
+            LOCAL,
+            additionalPredicate,
+            soupName,
+            idField
+        )
+    }
+
+    /**
+     * Fetches remote IDs still present on the server from the list of local IDs.
+     *
+     * @param syncManager SyncManager instance.
+     * @return List of IDs still present on the server.
+     */
+    @Throws(IOException::class, JSONException::class)
+    protected abstract fun getRemoteIds(
+        syncManager: SyncManager,
+        localIds: Set<String>
+    ): Set<String>
+
+    /**
+     * Gets the latest modification timestamp from the array of records.
+     * @param records
+     * @return latest modification time stamp
+     * @throws JSONException
+     */
+    @Throws(JSONException::class)
+    open fun getLatestModificationTimeStamp(records: JSONArray): Long {
+        return getLatestModificationTimeStamp(records, modificationDateFieldName)
+    }
+
+    /**
+     * When sync down fetches records from older to newer, the maxTimeStamp for the sync
+     * can be updated throughout the sync, and as a result running a paused (or killed) sync
+     * does not refetch all records.
+     * @return true if sync down is sorted by latest modification time stamp
+     */
+    open val isSyncDownSortedByLatestModification: Boolean
```

Comment:
Optional: It doesn't look like this needs a get accessor since it could simply `= false`.

Reply:
Fixed with https://github.com/forcedotcom/SalesforceMobileSDK-Android/pull/2459/commits/89de87833440f6973962d29e930ee88a909ec61b